osutil/io.go

```diff
@@ -41,16 +43,50 @@ const (
 // all unit tests in genreal.
 var snapdUnsafeIO bool = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test") && GetenvBool("SNAPD_UNSAFE_IO")
 
-// AtomicWriteFile updates the filename atomically and works otherwise
-// like io/ioutil.WriteFile()
-//
-// Note that it won't follow symlinks and will replace existing symlinks
-// with the real file
-func AtomicWriteFile(filename string, data []byte, perm os.FileMode, flags AtomicWriteFlags) (err error) {
-	return AtomicWriteFileChown(filename, data, perm, flags, -1, -1)
+// An AtomicWriter is an io.WriteCloser that has a Finalize() method that does
+// whatever needs to be done so the edition is "atomic": an AtomicWriter will
+// do its best to leave either the previous content or the new content in
+// permanent storage. It also has a Cancel() method to abort and clean up.
+type AtomicWriter interface {
+	io.WriteCloser
+
+	// Finalize the writing operation and make it permanent.
+	//
+	// If Finalize succeeds, the file is closed and further attempts to write will
+	// fail. If Finalize fails, Cancel() needs to be called to clean up.
+	Finalize() error
+
+	// Cancel closes the AtomicWriter, and cleans up any artifacts. Cancel
+	// can fail if Finalize() was (even partially) successful.
+	//
+	// It's an error to call Cancel once Close has been called.
+	Cancel() error
 }
 
-func AtomicWriteFileChown(filename string, data []byte, perm os.FileMode, flags AtomicWriteFlags, uid, gid int) (err error) {
+type atomicFile struct {
+	*os.File
+
+	target  string
+	tmpname string
+	uid     int
+	gid     int
+	renamed bool
+}
+
+// NewAtomicFile builds an AtomicWriter backed by an *os.File that will have
+// the given filename, permissions and uid/gid when Finalized.
+//
+//   It _might_ be implemented using O_TMPFILE (see open(2)).
```

Comment:
this is not done here, right?

Reply:
no, but I'll be doing it opportunistically; we still need the current code as `O_TMPFILE` is filesystem- and kernel-dependent, but the ability to use it has informed the latest iteration of the API (kudos to zyga for pointing out `O_TMPFILE`)